UOP1_Project/Assets/Scripts/Characters/WalkingParticle.cs

```diff
@@ -0,0 +1,37 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class WalkingParticle : MonoBehaviour
+{
+
+    [Tooltip("Particle prefab to be attached to the charactor in the scene.")] public ParticleSystem walkingParticlePrefab;
+
+    private Vector3 lastPosition;
+    private Vector3 currentPosition;
+    private ParticleSystem particleObject;
+
+    private void Update()
+    {
+        if (lastPosition == null) lastPosition = transform.position;
+        if (currentPosition == null) currentPosition = transform.position;
```

Comment:
If you check a struct against `null` it will always be false.
Either make it nullable or check it against `default(Vector3)` (that is `Vector3.zero` in this case). The later could introduce a bug, where the position really is `default`.

Reply:
More importantly - what is the check actually for?

There's a bunch of code here which looks like it is trying to do "set up on the first run" things - assigning the positions, instantiating the prefab, etc. It would be much better, and much simpler, to do that in an OnEnable() method, instead of doing it in Update().